From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Thu, 15 Jan 2026 00:16:29 +0300
Subject: [PATCH] Pufferfish: Optimize random calls in chunk ticking

This patch is based on the following patch:
"Optimize random calls in chunk ticking"
By: Paul Sauve <paul@technove.co>
As part of: Airplane (https://github.com/TECHNOVE/Airplane)
Licensed under: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index bdc92e717170a1502c7b8e2494b6cf48635501bf..369150dd4f3485fe1a18ceedf6e85084154f95ef 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1146,7 +1146,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         int minZ = chunkPos.getMinBlockZ();
         ProfilerFiller profiler = Profiler.get();
         profiler.push("thunder");
-        if (!this.paperConfig().environment.disableThunder && raining && this.isThundering() && this.spigotConfig.thunderChance > 0 && this.random.nextInt(this.spigotConfig.thunderChance) == 0) { // Spigot // Paper - Option to disable thunder
+        if (!this.paperConfig().environment.disableThunder && raining && this.isThundering() && this.spigotConfig.thunderChance > 0 && chunk.shouldDoLightning(this.random)) { // Spigot // Paper - Option to disable thunder // DivineMC - Pufferfish: Optimize random calls in chunk ticking
             BlockPos pos = this.findLightningTargetAround(this.getBlockRandomPos(minX, 0, minZ, 15));
             if (this.isRainingAt(pos)) {
                 DifficultyInstance difficulty = this.getCurrentDifficultyAt(pos);
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 2a2df7c210c60715fd33439efc494defe2d98550..e7e7f3036d9735dea3764478b265f8e5803b23be 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -139,6 +139,18 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         return this.getFluidState(x, y, z);
     }
     // Paper end - get block chunk optimisation
+    // DivineMC start - Pufferfish: Optimize random calls in chunk ticking
+    private int lightningTick;
+
+    public final boolean shouldDoLightning(net.minecraft.util.RandomSource random) {
+        if (this.lightningTick-- <= 0) {
+            this.lightningTick = random.nextInt(this.level.spigotConfig.thunderChance) << 1;
+            return true;
+        }
+
+        return false;
+    }
+    // DivineMC end - Pufferfish: Optimize random calls in chunk ticking
 
     public LevelChunk(final Level level, final ChunkPos pos) {
         this(level, pos, UpgradeData.EMPTY, new LevelChunkTicks<>(), new LevelChunkTicks<>(), 0L, null, null, null);
@@ -176,6 +188,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         this.debug = !empty && this.level.isDebug();
         this.defaultBlockState = empty ? VOID_AIR_BLOCKSTATE : AIR_BLOCKSTATE;
         // Paper end - get block chunk optimisation
+        this.lightningTick = new java.util.Random().nextInt(100000) << 1; // DivineMC - Pufferfish: Optimize random calls in chunk ticking
     }
 
     public LevelChunk(final ServerLevel level, final ProtoChunk protoChunk, final LevelChunk.@Nullable PostLoadProcessor postLoad) {
